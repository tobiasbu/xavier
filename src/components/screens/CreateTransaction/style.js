import { createUseStyles } from 'react-jss';

const useStyle = createUseStyles({
  registerButton: {
    width: 240,
    float: 'right',
  },
});

export default useStyle;
