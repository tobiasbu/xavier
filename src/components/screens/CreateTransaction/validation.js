import { splitCurrency } from '@utils/stringUtils';

/**
 * Currency validation
 * @param {string} current
 */
export function validateCurrency(current) {
  const r = splitCurrency(current);
  if (r.denominator === 0 && r.numerator === 0) {
    return '"Zero" reais não existe :P. Por favor insira um valor válido.';
  }
  return true;
}

/**
 * Description validation
 * @param {string} desc
 */
export function validateDescription(desc) {
  const len = desc.length;
  if (len === 1) {
    return 'Hum. Descreva melhor sua transação.';
  }
  return true;
}

/**
 * Check form errors.
 * @param {Object} error Error object.
 * @param {string} defaultErrorMessage Default error message.
 */
export function checkError(error, defaultErrorMessage = '') {
  if (!error) {
    return {
      message: '',
      valid: null,
    };
  }
  let msg = error.message;
  msg = msg.length > 0 ? msg : defaultErrorMessage;
  return {
    valid: false,
    message: msg,
  };
}
