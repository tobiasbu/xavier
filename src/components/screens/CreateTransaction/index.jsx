/* eslint-disable react/prefer-stateless-function */
import React, { useState } from 'react';
import useForm from 'react-hook-form';
import { addTransaction } from '@API';

import Button from '../../basic/Button';
import Input from '../../basic/Input';
import CurrencyInput from '../../basic/CurrencyInput';
import Toggle from '../../basic/Toggle';
import Header from '../../container/Header';
import Content from '../../container/Content';

import useStyle from './style';
import * as validation from './validation';

/**
 * Create transaction screen.
 *
 * Form based on the react-form hook.
 * @see https://blog.logrocket.com/react-hook-form-vs-formik-a-technical-and-performance-comparison/
 */
const CreateTransaction = () => {
  const {
    register, handleSubmit, errors, formState,
  } = useForm();
  const classes = useStyle();
  const [isDebit, setDebit] = useState(false);

  // On form submit
  const onSubmit = (data, e) => {
    e.preventDefault();
    addTransaction(data.valor, data.descricao, data['credito-debito']);
    e.target.reset();
  };

  const descError = validation.checkError(errors.descricao, 'Por favor, insira uma descrição para sua transação.');
  const currencyError = validation.checkError(errors.valor, 'Por favor, insira uma valor válido.');
  const buttonColor = (isDebit) ? 'neptune' : 'mars';

  return (
    <Content>
      <Header>Cadastre uma nova transação</Header>
      <main>
        <form onSubmit={handleSubmit(onSubmit)}>
          <p>Preencha os campos a seguir:</p>
          <Input
            label="Descrição"
            placeholder="Ex: Compras do Atacado, Manunteção da Cozinha, Fatura do Banco..."
            forwardedRef={
              register({
                required: true,
                validate: validation.validateDescription,
                min: 2,
              })
            }
            validation={descError.valid}
            errorMessage={descError.message}
          />
          <CurrencyInput
            label="Valor"
            forwardedRef={
              register({
                required: true,
                validate: validation.validateCurrency,
              })
            }
            validation={currencyError.valid}
            errorMessage={currencyError.message}
          />
          <Toggle
            label="Crédito"
            secondaryLabel="Débito"
            className="field-full"
            forwardedRef={register}
            onChange={(checked) => { setDebit(checked); }}
            title="Crédito ou Débito?"
          />
          <Button
            color={buttonColor}
            className={classes.registerButton}
            type="submit"
            disabled={formState.isSubmitting}
            title="Cadastrar nova transação"
          >
            Cadastrar
          </Button>
        </form>
      </main>
    </Content>
  );
};

export default CreateTransaction;
