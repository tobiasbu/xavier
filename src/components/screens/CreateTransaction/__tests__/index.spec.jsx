import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import CreateTransaction from '..';

describe('Create Transaction screen', () => {
  it('should render', () => {
    const component = shallow(
      <CreateTransaction />,
    );
    expect(toJSON(component)).toMatchSnapshot();
  });
});
