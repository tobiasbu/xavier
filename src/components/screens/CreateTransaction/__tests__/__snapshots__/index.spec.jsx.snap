// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Transaction screen should render 1`] = `
<Content>
  <Header>
    Cadastre uma nova transação
  </Header>
  <main>
    <form
      onSubmit={[Function]}
    >
      <p>
        Preencha os campos a seguir:
      </p>
      <Input
        defaultValue=""
        disabled={false}
        errorMessage=""
        forwardedRef={[Function]}
        label="Descrição"
        onChange={null}
        onInput={null}
        placeholder="Ex: Compras do Atacado, Manunteção da Cozinha, Fatura do Banco..."
        required={false}
        validation={null}
      />
      <CurrencyInput
        defaultValue=""
        disabled={false}
        errorMessage=""
        forwardedRef={[Function]}
        label="Valor"
        onBlur={null}
        onChange={null}
        onControlClick={null}
        onInput={null}
        placeholder={null}
        step={10}
        validation={null}
      />
      <Toggle
        checked={false}
        className="field-full"
        disabled={false}
        forwardedRef={[Function]}
        label="Crédito"
        onChange={[Function]}
        secondaryLabel="Débito"
        title="Crédito ou Débito?"
      />
      <Button
        className="registerButton-0-2-1"
        color="mars"
        disabled={false}
        title="Cadastrar nova transação"
        type="submit"
      >
        Cadastrar
      </Button>
    </form>
  </main>
</Content>
`;
